docs/04-reference/winglang-spec.md

```diff
@@ -869,30 +869,234 @@ let f = (arg1: num, var arg2: num) => {
+
+```js
+struct FormatOpts {
+  radix: num = 10;
+  some_optional: str?;
+}
+
+let opts = FormatOpts {};
+assert(opts.radix == 10);
+assert(opts.some_optional? == false); // <-- no value inside `some_optional`
+```
+
+A value can be omitted from a struct literal if the field is optional _or_ if it has a default value in the struct
+declaration. If the field is optional, it's type is `T?`. If it has a default value it's type is `T` (since we know it will always have a value).
+In the above example the type of `radix` is `num`.
+
+##### 1.7.1.2 Declaration in variables
+
+Use `T?` to indicate that a variable is optional:
+
+```js
+let var y: num?;
+assert(y? == false); // y doesn't have a value
+
+// ok to reassign another value because `y` is reassignable (`var`)
+y = 123;
+assert(y? == true);
+```
+
+It is okay to initialize a value:
+
+```js
+let var y: num? = 12;
+```
+
+##### 1.7.1.3 Declaration in class fields
+
+Similarly to struct fields, fields of classes can be also defined as optional using `T?`:
+
+```js
+class Foo {
+  my_opt: num?;
+  var my_var: str?;
+
+  init(opt: num?) {
+    this.my_opt = opt;
+  }
+
+  set_my_var(x: str) {
+    this.my_var = x;
+  }
+}
+```
+
+##### 1.7.1.4 Declaration in function arguments
+
+In the following example, the argument `by` is optional, so it is possible to call `x()` without
+supplying a value for `by`:
+
+```js
+let increment = (x: num, by: num?): num => {
+  return x + (by ?? 1);
+};
+
+assert(my_function(88) == 89);
+assert(my_function(88, 2) == 90);
+```
+
+Alternatively, specifying a default value will also make an argument optional. Using a default value
+in the function declaration ensures that `by` always has a value so there is no need to unwrap it
+(this is why its type is `num` and not `num?`):
+
+```js
+let increment = (x: num, by: num = 1): num {
+  return x + by;
+}
+```
+
+Non-optional arguments can only be used before all optional arguments:
+
+```js
+let my_fun = (a: str, x?: num, y: str): void = { /* ... */ };
+//-----------------------------^^^^^^ ERROR: cannot declare a non-optional argument after an optional
+```
+
+If a function uses a keyword argument struct as the last argument, and there are other optional arguments before,
+it also has to be declared as optional:
+
+```js
+let parse_int = (x: str, radix: num?, opts?: ParseOpts): num { /* ... */ };
+// or
+let parse_int = (x: str, radix: num = 10, opts: ParseOpts = ParseOpts {}): num { /* ... */ };
+```
+
+##### 1.7.1.5 Declaration in function return types
+
+If a function returns an optional type, use the `return nil;` statement to indicate that the value is
+not defined.
+
+> This is currently the only scenario in the language where `nil` is allowed.
```

Comment:
I believe this is conflicted with an example in section 1.8
```
let i1? = nil;
```

Reply:
Fixed